l10n_ch: QR Invoice v2

yvaucher opened the pull request:
- [x] Depends on #35873 
- [ ] for tests fixing depends on #35979 

[this needs rebase]


## Description of the issue/feature this PR addresses:


The QR-Invoice is planned to land in June 2020. (as it was delayed)
Since it was developed

[IMP] l10n_ch: Swiss QR Invoice version 0200

Swiss QR Invoice with QR Reference (QRR) needs a special
IBAN named QR-IBAN.
A QRR is exactly the same as an ISR Reference.

* Add a new res.partner.bank type `qr-iban`
* Fix Reference on QR-invoice which was subscription number instead of QRR
* Move QR code logic for report on invoice instead of res.partner.bank
* Change addresses to "K", combined, in order to not have to split the building number from the addresses
* Upgrade QR-Invoice to 0200

reminder: *to activate Swiss QR-code add the system parameter* **l10n_ch.print_qrcode**

https://www.paymentstandards.ch/dam/downloads/ig-qr-bill-en.pdf

## Current behaviour before PR:

The version was still version 1.
The reference used was a the subscription number (previously badly named ISR reference on the bank).

## Desired behaviour after PR is merged:

Version 2 of QR-Invoice specs.

The reference on the invoice QR must be a Payment Reference in our case the ISR reference which is computed on the invoice (`l10n_ch_isr_number`)

### Current output with fixed reference

![2019-08-21-182906_1064x709_scrot](https://user-images.githubusercontent.com/4158438/63450265-ec6d1180-c441-11e9-9c91-4d9bb6a63422.png)



To be done later, add the optional bill information from Swico standards.


Be able to print a QR code with the ISR reference of the invoice (account.move).



--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

yvaucher commented:
@jco-odoo  @hpr-odoo here is an other PR to fix the QR Invoice, not urgent as it will only land in 2020. But I need some help to fix the CacheMiss error. Seems it's new in Odoo 13. Any idea how to fix them, and what could cause those?

jco-odoo commented:
@yvaucher  It is because in the compute you are required to give a value for all the records now. (so if you do a filtered, you still need to give a value for those you did not set)  I am also on it for my other branches.

yvaucher commented:
@jco-odoo thanks, I fixed the computed fields in https://github.com/odoo/odoo/pull/35979

(PR that fixes the onchange on res.partner.bank acc_number field.)

yvaucher commented:
I rebased on top of #35979 and added a test on swissqr generation.

RIP to the old test on splitting num from street.
It's way easier to not try to get the number and provide the address lines.
By the way there are some address without numbers.

Example:

> Camptocamp SA
> 
> EPFL Innovation Park
> Building A
> 1015 Lausanne

oco-odoo commented:
@yvaucher Thanks for sharing this with me. As I told you, I'm currently refactoring the whole QR-code generation stuff (not only for Switzerland, but I impact it as well). It's still being developed, and there is no PR yet, but I will let you know once it's done (branch is called master-qr-code-refact-oco , but it's still a bit full of TODOs for now). I'll take inspiration in your code to clean the implementation of Swiss QR-codes. I'm closing this PR, but it does not mean it was useless ;)

yvaucher commented:
Superseeded by https://github.com/odoo/odoo/pull/44839
